### DISCUSSION TITLE
[Question] How to show loading screen on intermediate authstate

### DISCUSSION THREAD

**Discussion by kornha** [csells/go_router#211] [Q&A] [answered]
Killer package, thanks so much once again. I wanted to run by an important use case here for myself, and I imagine for others, and hopefully get your input.

If I follow the sample, I get something like this:
```
   redirect: (state) {
          final loggedIn = _authState.isLoggedIn;
          final authPath = state.location != "/login" && state.location != "/";
          final loadingPath = state.location == "/loading";
          
          // the user is not logged in and not headed to /login, they need to login
          if (!loggedIn && authPath) return '/login';

          // the user is logged in and headed to /login, no need to login again
          if (loggedIn && !authPath) {
            return "/home";
          }

          // no need to redirect at all
          return null;
        },
```
This works great, the only problem is, if I refresh the page the authstate is momentarily interdeterminate. I can use this state and assume the user is logged out, which the above code down, but then for a hot second it shows the login screen. I would rather it show a **loading** screen in this intermediate auth state.

There's 2 approaches I can think of to address this use case.
1: in my actual login view, show a loading screen during an intermediate authState.
This does not work, as then, for a very hot second, the screen will update to the login screen before the router redirects to the authenticated page
2: A bit hacky, but to do this on the router level, something like:
```
        redirect: (state) {
          final loggedIn = _authState.isLoggedIn;
          final loading = _authState.isLoading;
          final loadingPath = state.location == "/loading";
          final authPath = state.location != "/login" &&
              state.location != "/" &&
              !loadingPath;

          if (loading && loadingPath) return null;

          if (loading) return "/loading";

          if (!loggedIn && (authPath || loadingPath)) return "/login";

          if (loggedIn && (!authPath || loadingPath)) return "/home";

          return null;
        },
```
This works perfectly from a UX standpoint, but has the minor problem that it includes an intermediate "loading" in the auth path, and just feels hacky.
Is this the preferred way to solve this problem?

**Comment by csells** [comment]
Why is the auth state indeterminate? Can you post a minimal sample that shows the login page for a "hot second"?

**Comment by kornha** [comment]
Sure. The authState will be intermediate because the server has yet to respond with authenticated or unauthenticated.
For example, here is a changenotifier that this would exist in:

```
class AuthState extends ChangeNotifier {
  AuthStatus status = AuthStatus.unknown;
  final Auth _auth = Auth();
  User? authUser;

  bool get isLoggedIn => status == AuthStatus.authenticated;
  bool get isLoading =>
      status == AuthStatus.unknown || status == AuthStatus.authenticating;

  AuthState.instance() {
    status = AuthStatus.unknown;
    _auth.auth.authStateChanges().listen((user) async {
      authUser = user;
      if (user == null) {
        status = AuthStatus.unauthenticated;
      } else {
        status = AuthStatus.authenticated;
      }
      notifyListeners();
    });
  }
}
```
Note that, in the above, before a server response we will have state == unknown.

**Comment by kornha** [comment]
The problem with showing the login page with a loading indicator -> redirecting to an authenticated page, is that it creates a race condition. The loading indicator and the redirect will both listen on a change in authState. So if the loading indicator hears the change first, it will vanish, then show the login page (for an instant.. so I guess I could do a 'sleep' here but certainly not ideal), then redirect to the authenticated page. Does that make sense? I can image capture if not.

**Comment by kornha** [comment]
A subtle annoyance with my proposed solution as well is that it requires deeplinking to be implemented in the loading path.

**Comment by lulupointu** [comment]
May this help ? (I used `null` for `AuthState.unknown` for simplicity)

<details>
<summary>
Click to show code
</summary>

```dart
import 'package:flutter/material.dart';
import 'package:go_router/go_router.dart';

void main() => runApp(BooksApp());

class Credentials {
  final String username;
  final String password;

  Credentials(this.username, this.password);
}

abstract class Authentication {
  Future<bool> isSignedIn();

  Future<void> signOut();

  Future<bool> signIn(String username, String password);
}

class MockAuthentication implements Authentication {
  bool _signedIn = false;

  @override
  Future<bool> isSignedIn() async {
    return _signedIn;
  }

  @override
  Future<void> signOut() async {
    _signedIn = false;
  }

  @override
  Future<bool> signIn(String username, String password) async {
    return _signedIn = true;
  }
}

class AuthState extends ChangeNotifier {
  final Authentication _auth;
  bool? _isSignedIn = null;

  AuthState({required Authentication auth}) : _auth = auth;

  Future<void> initialize() async {
    await Future<void>.delayed(Duration(seconds: 2));
    _isSignedIn = false;
    notifyListeners();
  }

  Future<bool> signIn(Credentials credentials) async {
    var success = await _auth.signIn(credentials.username, credentials.password);
    _isSignedIn = success;
    notifyListeners();
    return success;
  }

  Future<void> signOut() async {
    await _auth.signOut();
    _isSignedIn = false;
    notifyListeners();
  }

  bool? get isSignedIn => _isSignedIn;
}

class BooksApp extends StatefulWidget {
  @override
  State<BooksApp> createState() => _BooksAppState();
}

class _BooksAppState extends State<BooksApp> {
  final authState = AuthState(auth: MockAuthentication());

  @override
  void initState() {
    authState.initialize();
    super.initState();
  }

  @override
  Widget build(BuildContext context) => MaterialApp.router(
        routeInformationParser: _router.routeInformationParser,
        routerDelegate: _router.routerDelegate,
        title: 'GoRouter Example: LogIn',
      );

  late final _router = GoRouter(
    navigatorBuilder: (_, child) => AuthStateUpdateHandler(
      child: child!,
      authState: authState,
    ),
    redirect: (state) {
      final isSignedIn = authState.isSignedIn;
      if (isSignedIn == null && state.location != '/loading') {
        return '/loading';
      }
      if (isSignedIn == false && state.location != '/sign-in') {
        return '/sign-in';
      }
      if (isSignedIn == true &&
          (state.location == '/loading' || state.location == '/sign-in')) {
        return '/';
      }
    },
    routes: [
      GoRoute(
        path: '/',
        pageBuilder: (_, state) => MaterialPage<void>(
          key: state.pageKey,
          child: HomeScreen(onSignOut: authState.signOut),
        ),
        routes: [
          GoRoute(
            path: 'books',
            pageBuilder: (_, state) => MaterialPage<void>(
              key: state.pageKey,
              child: BooksListScreen(),
            ),
          ),
        ],
      ),
      GoRoute(
        path: '/sign-in',
        pageBuilder: (_, state) => MaterialPage<void>(
          key: state.pageKey,
          child: SignInScreen(onSignedIn: authState.signIn),
        ),
      ),
      GoRoute(
        path: '/loading',
        pageBuilder: (_, state) => MaterialPage<void>(
          key: state.pageKey,
          child: Center(child: Text('Loading signin state')),
        ),
      ),
    ],
    errorPageBuilder: (context, state) => MaterialPage<void>(
      key: state.pageKey,
      child: Text('${state.error}'),
    ),
  );
}

class AuthStateUpdateHandler extends StatefulWidget {
  final Widget child;
  final AuthState authState;

  AuthStateUpdateHandler({required this.child, required this.authState});

  @override
  State<AuthStateUpdateHandler> createState() => _AuthStateUpdateHandlerState();
}

class _AuthStateUpdateHandlerState extends State<AuthStateUpdateHandler> {
  void _onAuthStateChange() {
    final isSignedIn = widget.authState.isSignedIn;
    if (isSignedIn == null) return context.go('/loading');
    context.go(isSignedIn ? '/' : '/log-in');
  }

  @override
  void initState() {
    super.initState();
    widget.authState.addListener(_onAuthStateChange);
  }

  @override
  void dispose() {
    widget.authState.removeListener(_onAuthStateChange);
    super.dispose();
  }

  @override
  Widget build(BuildContext context) => widget.child;
}

class HomeScreen extends StatelessWidget {
  final VoidCallback onSignOut;

  HomeScreen({required this.onSignOut});

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(),
      body: Center(
        child: Column(
          children: [
            ElevatedButton(
              onPressed: () => context.go('/books'),
              child: Text('View my bookshelf'),
            ),
            ElevatedButton(
              onPressed: onSignOut,
              child: Text('Sign out'),
            ),
          ],
        ),
      ),
    );
  }
}

class SignInScreen extends StatefulWidget {
  final ValueChanged<Credentials> onSignedIn;

  SignInScreen({required this.onSignedIn});

  @override
  _SignInScreenState createState() => _SignInScreenState();
}

class _SignInScreenState extends State<SignInScreen> {
  String _username = '';
  String _password = '';

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(),
      body: Center(
        child: Column(
          children: [
            TextField(
              decoration: InputDecoration(hintText: 'username (any)'),
              onChanged: (s) => _username = s,
            ),
            TextField(
              decoration: InputDecoration(hintText: 'password (any)'),
              obscureText: true,
              onChanged: (s) => _password = s,
            ),
            ElevatedButton(
              onPressed: () => widget.onSignedIn(Credentials(_username, _password)),
              child: Text('Sign in'),
            ),
          ],
        ),
      ),
    );
  }
}

class BooksListScreen extends StatelessWidget {
  BooksListScreen();

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(),
      body: ListView(
        children: [
          ListTile(
            title: Text('Stranger in a Strange Land'),
            subtitle: Text('Robert A. Heinlein'),
          ),
          ListTile(
            title: Text('Foundation'),
            subtitle: Text('Isaac Asimov'),
          ),
          ListTile(
            title: Text('Fahrenheit 451'),
            subtitle: Text('Ray Bradbury'),
          ),
        ],
      ),
    );
  }
}
```

</details>

This could be improved of course but I think the core idea is there

**Comment by csells** [comment]
I'm sorry. I'm looking for a complete minimal repro, ie an app that I can load into my ide and run.

**Comment by lulupointu** [comment]
Oh actually I was giving a solution to @kornha ^^

**Comment by csells** [comment]
Ha. And I was asking @kornha for a repro so I could understand the problem. One step ahead as usual @lulupointu. : )

**Reply by kornha** [reply]
Here is an example screen capture of the race condition mentioned above:
![race_condition](https://user-images.githubusercontent.com/5386694/144739566-79bcbe4a-e608-42e2-ae64-07f25abcd9d7.gif)

**Comment by kornha** [comment]
@lulupointu Thanks for the solution! It is similar to my proposed solution, but I have a few questions on it

1. Won't you have to implement deeplinking for you loading page?
2. I don't understand why we need the high-level redirect, and also do some route manipulation here:
```
    final isSignedIn = widget.authState.isSignedIn;
    if (isSignedIn == null) return context.go('/loading');
    context.go(isSignedIn ? '/' : '/log-in');
```
Can you not just use the **refreshListenable** feature of the router?
@csells I wonder if it would be cool to have a loadingState builder for something like this? Just a thought to consider, you're already killing it with the repo!!!!

**Comment by lulupointu** [comment]
> 1. Won't you have to implement deeplinking for you loading page?

I don't understand what you mean

> 2. I don't understand why we need the high-level redirect, and also do some route manipulation here (...). 
> Can you not just use the refreshListenable feature of the router?

I haven't tried but I don't think so. 
For example: If you are NOT signed in; and you deep-link (while your app is already running, e.g. by typing a url in the browser) to `'/'` then I don't think **refreshListenable** would be triggered but you still would have to redirect.
Then I don't have any experience with **refreshListenable**, I fond the idea cool but in real application I would always use the method I described in my example (for the reason described here).

**Reply by kornha** [reply]
For 1. Let's say you have both a "sign-in" and "landing" page, neither of them require authentication. When you refresh the page, your authState will be momentarily null, and route to the loading screen. Following which, you will then route to your "sign-in" page, even if you initially wanted to hit the landing page.

**Reply by kornha** [reply]
Ergo, if you want to go straight to your Landing page (and not sign-in page), you will require the loading page to deeplink the landing page

**Reply by kornha** [reply]
For 2 I think you'd be covered with the refreshListenable :) either way no big deal, to each their own

**Comment by lulupointu** [comment]
> For 1. Let's say you have both a "sign-in" and "landing" page, neither of them require authentication. When you refresh the page, your authState will be momentarily null, and route to the loading screen. Following which, you will then route to your "sign-in" page, even if you initially wanted to hit the landing page

Yeah that's what I meant by *"This could be improved of course but I think the core idea is there"*. In reality whenever you redirect to `loading` I would add a `queryParameter` named `from` for example, and redirect to this instead.

> For 2 I think you'd be covered with the refreshListenable :) either way no big deal, to each their own

Do test it if you choose this since I have some doubt. If you want to share your result with a small snippet (for example a modified version of mine) this could be great for the future person looking here.

**Reply by kornha** [reply]
> Yeah that's what I meant by "This could be improved of course but I think the core idea is there". In reality whenever you redirect to loading I would add a queryParameter named from for example, and redirect to this instead.

This is deeplinking, no? :)

**Reply by lulupointu** [reply]
Oh idk I though deep-linking was trying to get in the app with something else than the default initial path. Anyway I prefer stating explicitly what I mean so that anyone can understand ^^

**Comment by kornha** [comment, marked as answer]
My current solution, tested and working for me:

```
        refreshListenable: _authState,
        redirect: (state) {
          final loggedIn = _authState.isLoggedIn;
          final loading = _authState.isLoading;
          final loadingPath = state.location.startsWith("/loading");
          final authPath = state.location != "/login" &&
              state.location != "/" &&
              !loadingPath;

          if (loading && loadingPath) return null;

          if (loading) return "/loading?from=${state.subloc}";

          if (!loggedIn && (authPath || loadingPath)) {
            return state.queryParams['from'] ?? "/";
          }

          if (loggedIn && (!authPath || loadingPath)) {
            return state.queryParams['from'] ?? "/home";
          }

          return null;
        },
```

**Reply by lulupointu** [reply]
Oh I see, I thought you meant to remove the `redirect` part as well.

Alright then this should work indeed, and is simpler than my solution :wink:
